### PR TITLE
feat (CodeCoverage): use karma-remap-istanbul

### PR DESCRIPTION
**Pull request by deebloo**
use karma-remap-istanbul to map code coverage back to source typescript file

### CODE REVIEW

**Comment by deebloo** [comment]
@Brocco I am seeing the CI builds have been running for like 20 min. I don't believe I would have done anything here to cause this?

**Comment by filipesilva** [comment]
Heya! Sometimes the builds just take a while. In travis some show has having errored though.

I'm not too sure of the purpose of this PR though, can you tell me what's the issue you're trying to solve?

**Comment by deebloo** [comment]
Sure! So currently. The code coverage report displays the compiled js files and not the source ts files. This should fix that

**Comment by filipesilva** [comment]
Ok I understand it now. Yeap, that needs fixing, thank you for taking it up. Me and @TheLarkInn should be reviewing when the CI is stable, since we worked on the current implementation.

**Comment by deebloo** [comment]
👍

**Comment by delasteve** [comment]
As a maintainer (sure?) of `karma-remap-istanbul`, I would say this is not the way to go, currently. I got promoted to maintainer because I had a PR out and the owner couldn't find the time to merge my PR in. 

I've had issues with the plugin, but don't know enough about karma plugins to be considered a maintainer. If someone would like to take a look at it, by all means, please do. There are two PRs on the repo that use `karma-coverage` as a base. I like the implementation, but there are still concerns it's not all the way correct.

@devCrossNet was working on a fix that created a task that used `remap-istanbul` instead of using the flaky plugin. I still recommend this is the way to go, since the current implementation of `karma-remap-istanbul` needs work.

**Comment by devCrossNet** [comment]
@delasteve Hello, it's me ... ;-) I could update my coverage branch to keep my implementation up to date with the new webpack stuff. But I am still waiting for the remap-istanbul release. 

Or should I stop working on this feature?

**Comment by devCrossNet** [comment]
I had a short look at the new implementation. the test process changed and there is some new kind of karma plugin. I need some time to dive into the Code and get everything working again.

**Comment by filipesilva** [comment]
The new karma plugin was added to have the whole build process for tests managed via karma. I made it but I wouldn't say I'm very knowledgeable of karma plugins, it's just editing the config really.

**Comment by filipesilva** [comment]
@deebloo tests seem to be failing a lot with this PR, I can restart the build but the instability seems due the the changes introduced in the PR.

File: addon/ng2/blueprints/ng2/files/config/karma.conf.js [hunk 1]
```diff
@@ -22,10 +23,34 @@ module.exports = function (config) {
       { pattern: './src/test.ts', watched: false }
     ],
     preprocessors: {
+      './src/**/**/*.ts': ['angular-cli-coverage'],
```
Comment (TheLarkInn) [inline addon/ng2/blueprints/ng2/files/config/karma.conf.js]: Would you be willing to explain this?

**Comment by devCrossNet** [comment]
@filipesilva  I understand that code. It is not much,I had trouble with the source maps and the processing of code coverage. But I tried it just a few minutes. I'll come back to you when I can't get it working again after a few hours.

File: addon/ng2/blueprints/ng2/files/config/karma.conf.js (hunk 1 above)
Comment (deebloo) [inline addon/ng2/blueprints/ng2/files/config/karma.conf.js]: I will look at it today. and try to find out exactly why that is.

Comment (deebloo) [inline addon/ng2/blueprints/ng2/files/config/karma.conf.js]: I added a new preprocessor in the karma plugin. While there is a single entry point for the tests in order to get the remap plugin to generate the mapped coverage report it needs to  run over the test files themselves.

**Comment by deebloo** [comment]
@filipesilva I am running my local tests again to see if I can spot what is wrong. Also, based on the conversation here do we want to try NOT using the remap karma plugin?

**Comment by TheLarkInn** [comment]
cc @bryanforbes do you think you could lend your coverage and instrumentation expertise. Id love to make coverage as accurate as possible and you've worked on this extensively also.

**Comment by filipesilva** [comment]
@deebloo I can't say one way or the other since I don't fully understand the question that is being brought up, but I have a considerable amount of respect for @delasteve and @devCrossNet since they put a lot of work into https://github.com/angular/angular-cli/pull/895 (which I've closed with when changing to webpack) so I'd like to understand what a better approach would be.

**Comment by deebloo** [comment]
@filipesilva fair. I will at least try to fix what I have so that this PR doesn't fail the tests, but after that I turn it over to the more experienced folks in the group

**Comment by deebloo** [comment]
update. so the tests just seem to hang the second time the e2e tests run the unit tests a second time. I can't find exactly why that is. It may be an issue with the karam plugin for remap. If that is the case (and based on other feed back in this feed). maybe I should just close this PR for now and let some of the other people who had been working on it take over. Thoughts?

**Comment by devCrossNet** [comment]
I evaluated the code yesterday and I think my approach is not going to work with the new webpack code (there are no files after the karma run, so I can not access the sourcemap files). I'll figure out another solution during the week. maybe the way would be a self-written, CLI optimized karma plugin-in.

**Comment by bryanforbes** [comment]
There are a couple of issues at play that should probably be addressed for more accurate code coverage:

1. `karma-coverage` currently outputs the "preamble" for coverage numbers into the top of each module. What this means is that each module that coverage is needed for must be evaluated even if that module doesn't have tests run on it. For instance, let's say we have modules `a.js`, `b.js`, and `c.js`. We write tests for `a.js` and `b.js`, but not for `c.js`. But in order to get `c.js` into the coverage report, one has to loop over all of the modules in a context and evaluate them (`const modules = require.context("../app", true); modules.keys().forEach(modules);`). However this presents a problem because `c.js` should never be evaluated and should have 0% coverage; with the context loop, any module setup code will be evaluated and added to the coverage report as executed statements which will not give an accurate coverage number. Instead, the full preamble of all modules (or all pertinent modules) should be output to the bootstrap chunk and a piece of code to set up the variable that istanbul uses in the module should be output to the top of each module. I currently have a plugin to do just that, but I haven't had the time to submit it to `karma-coverage` as an enhancement.

2. `karma-remap-istanbul` currently watches for a JSON output by `karma-coverage` which can lead to unnecessary errors during testing or to hangs while it waits for a file or both. I'm currently using a simple plugin that takes the `coverage` property off the results passed to `onBrowserComplete` by `karma` and runs it though `remap-istanbul` via the programmatic API. This also eliminates the need for `karma-coverage` to output an intermediate JSON file for `karma-remap-istanbul` to pick up and users will end up with just the coverage reports they specify (eliminating potential confusion).

3. The configuration passed to `awesome-typescript-loader` should not include any source map settings. Instead, the `devtool` property should be set to `inline-source-map` which will let webpack properly set up the source maps so `remap-istanbul` can read them.

I would be glad to submit PRs to both `karma-coverage` and `karma-remap-istanbul` to fix the first two issues, but I thought I'd weigh in on this at the request of @TheLarkInn. Let me know if there's any other way I can help.

**Comment by delasteve** [comment]
I merged in @bryanforbes's PR for `karma-remap-istanbul` and released it as `0.2.0`. Assuming this fixes everything with the karma plugin, it should take a quick update to get this in.

**Comment by deebloo** [comment]
I keep getting.  ```Uncaught TypeError: Cannot read property 'f' of undefined``` with the new plugin

**Comment by bryanforbes** [comment]
@deebloo What version of the plugin? @delasteve just released `0.2.1` which fixes a dumb mistake I made.

**Comment by deebloo** [comment]
@bryanforbes I am using 2.1. still getting that weird f error

**Comment by deebloo** [comment]
@bryanforbes I am getting this error now. Not sure if its my setup or the plugin
```
01 08 2016 18:09:19.292:ERROR [karma]: Error: Unable to find entry for [src/test.ts]
    at MemoryStore.Store.mix.get (/Users/dannyblue/Documents/projects/foo/node_modules/istanbul/lib/store/memory.js:38:19)
    at HtmlReport.Report.mix.writeDetailPage (/Users/dannyblue/Documents/projects/foo/node_modules/remap-istanbul/node_modules/istanbul/lib/report/html.js:411:67)
    at /Users/dannyblue/Documents/projects/foo/node_modules/remap-istanbul/node_modules/istanbul/lib/report/html.js:489:26
    at SyncFileWriter.extend.writeFile (/Users/dannyblue/Documents/projects/foo/node_modules/remap-istanbul/node_modules/istanbul/lib/util/file-writer.js:57:9)
    at FileWriter.extend.writeFile (/Users/dannyblue/Documents/projects/foo/node_modules/remap-istanbul/node_modules/istanbul/lib/util/file-writer.js:147:23)
    at /Users/dannyblue/Documents/projects/foo/node_modules/remap-istanbul/node_modules/istanbul/lib/report/html.js:488:24
    at Array.forEach (native)
    at HtmlReport.Report.mix.writeFiles (/Users/dannyblue/Documents/projects/foo/node_modules/remap-istanbul/node_modules/istanbul/lib/report/html.js:482:23)
    at /Users/dannyblue/Documents/projects/foo/node_modules/remap-istanbul/node_modules/istanbul/lib/report/html.js:484:22
    at Array.forEach (native)
    at HtmlReport.Report.mix.writeFiles (/Users/dannyblue/Documents/projects/foo/node_modules/remap-istanbul/node_modules/istanbul/lib/report/html.js:482:23)
    at HtmlReport.Report.mix.writeReport (/Users/dannyblue/Documents/projects/foo/node_modules/remap-istanbul/node_modules/istanbul/lib/report/html.js:566:14)
    at /Users/dannyblue/Documents/projects/foo/node_modules/remap-istanbul/lib/writeReport.js:77:22
    at /Users/dannyblue/Documents/projects/foo/node_modules/amdefine/amdefine.js:125:34
    at _combinedTickCallback (internal/process/next_tick.js:67:7)
    at process._tickCallback (internal/process/next_tick.js:98:9)
```

**Comment by bryanforbes** [comment]
I believe you'll need to pass the `force-sourcemaps` property to `sourcemap-istanbul-instrumenter-loader` in the [webpack config](https://github.com/angular/angular-cli/blob/master/addon/ng2/models/webpack-build-test.js#L68-L74) in order for that loader to produce a source map from the instrumented code to the original source:

```js
        {
          test: /\.(js|ts)$/, loader: 'sourcemap-istanbul-instrumenter-loader',
          exclude: [
            /\.(e2e|spec)\.ts$/,
            /node_modules/
          ],
          query: { 'force-sourcemaps': true }
        }
```

I'm in the process of trying to get a similar change made in `istanbul-instrumenter-loader`. If the author is unresponsive, I'll publish a loader I'm currently using which addresses point 1 above as well as source maps.

**Comment by deebloo** [comment]
should be working and passing all tests now.

**Comment by delasteve** [comment]
Thank you, @deebloo, @TheLarkInn, @bryanforbes, and @devCrossNet, for all the hard work you put into this. 

:+1: :tada:

**Comment by TheLarkInn** [comment]
@deebloo would you be willing to add one or two tests in our e2e workflow that can test this feature. 

This will allow us to safeguard this featrue against future changes. Otherwise @filipesilva LGTM.

**Comment by deebloo** [comment]
@TheLarkInn Ill add one after all of the "test tests" run to make sure the correct folders exist

**Comment by delasteve** [comment]
@deebloo This might help you. It's from my PR when I was implementing this.

https://github.com/angular/angular-cli/pull/895/files#diff-233c7c3510139cc00a622f09ce37097bR199

File: addon/ng2/blueprints/ng2/files/config/karma.conf.js [hunk 2]
```diff
@@ -24,8 +25,19 @@ module.exports = function (config) {
     preprocessors: {
       './src/test.ts': ['angular-cli']
     },
+    coverageReporter: {
+      dir: 'coverage',
+      reporters: [
+        { type : 'json', subdir: '.', file: 'coverage-final.json'}
+      ]
+    },
```
Comment (bryanforbes) [inline addon/ng2/blueprints/ng2/files/config/karma.conf.js]: Is there a reason to keep using `karma-coverage`? `karma-remap-istanbul` produces the coverage report and no longer needs the intermediate file and instrumenting is done by `sourcemap-istanbul-instrumenter-loader`.

Comment (deebloo) [inline addon/ng2/blueprints/ng2/files/config/karma.conf.js]: you are correct. I misunderstood the docs for the karma plugin

**Comment by TheLarkInn** [comment]
Thank you everyone for your work.

**Comment by filipesilva** [comment]
Just merged it in. A big thanks to @delasteve, @devCrossNet, @deebloo and @bryanforbes for making this work! It's been long in the backburner and the final solution is very elegant and clean imho.

**Comment by deebloo** [comment]
happy to help! you all are killing it with this newest update. (you know. in a good way).

**Pull request merged**